tests/Unit/Options/ServiceBasedMerchantStateTest.php

```diff
@@ -0,0 +1,130 @@
+
+	/** @var MockObject|OptionsInterface $options */
+	protected $options;
+
+	/** @var ServiceBasedMerchantState $service_based_merchant_state */
+	protected $service_based_merchant_state;
+
+	public function setUp(): void {
+		parent::setUp();
+
+		$this->options                      = $this->createMock( OptionsInterface::class );
+		$this->service_based_merchant_state = new ServiceBasedMerchantState();
+		$this->service_based_merchant_state->set_options_object( $this->options );
+	}
+
+	public function test_is_service_based_merchant_returns_cached_value_when_option_exists() {
+		$this->options->expects( $this->once() )
+			->method( 'get' )
+			->with( OptionsInterface::IS_SERVICE_BASED_MERCHANT )
+			->willReturn( true );
+
+		$result = $this->service_based_merchant_state->is_service_based_merchant();
+
+		$this->assertTrue( $result );
+	}
+
+	public function test_is_service_based_merchant_calculates_when_option_is_null() {
+		// Create a physical product that requires shipping
+		$physical_product = WC_Helper_Product::create_simple_product();
+		$physical_product->set_virtual( false );
+		$physical_product->save();
+
+		$this->options->expects( $this->once() )
+			->method( 'get' )
+			->with( OptionsInterface::IS_SERVICE_BASED_MERCHANT )
+			->willReturn( null );
+
+		// Should calculate and save when option is null
+		$this->options->expects( $this->once() )
+			->method( 'update' )
+			->with( OptionsInterface::IS_SERVICE_BASED_MERCHANT, false );
+
+		$result = $this->service_based_merchant_state->is_service_based_merchant();
+
+		// Has physical products, so NOT service-based
+		$this->assertFalse( $result );
+
+		// Cleanup
+		$physical_product->delete( true );
+	}
+
+	public function test_has_physical_products_returns_true_when_physical_products_exist() {
+		// Create a physical product that requires shipping
+		$physical_product = WC_Helper_Product::create_simple_product();
+		$physical_product->set_virtual( false );
+		$physical_product->save();
+
+		$result = $this->service_based_merchant_state->has_physical_products();
+
+		$this->assertTrue( $result );
+
+		// Cleanup
+		$physical_product->delete( true );
+	}
+
+	public function test_has_physical_products_returns_false_when_no_physical_products_exist() {
+		// Create a virtual product
+		$virtual_product = WC_Helper_Product::create_simple_product();
+		$virtual_product->set_virtual( true );
+		$virtual_product->save();
+
+		$result = $this->service_based_merchant_state->has_physical_products();
+
+		$this->assertFalse( $result );
+
+		// Cleanup
+		$virtual_product->delete( true );
+	}
+
+	public function test_has_physical_products_ignores_virtual_products() {
+		$virtual_product = WC_Helper_Product::create_simple_product();
+		$virtual_product->set_virtual( true );
+		$virtual_product->save();
+
+		$result = $this->service_based_merchant_state->has_physical_products();
+
+		$this->assertFalse( $result );
+
+		// Cleanup
+		$virtual_product->delete( true );
+	}
+
```

Comment:
Can we add at least one unit test to cover batch processing?

For example, create 150 products, then add a physical product at the end and verify that the batch processing behaves as expected. This would help ensure the logic works correctly for larger datasets and edge cases.